examples/js/pmrem/PMREMGenerator.js

```diff
@@ -1,293 +1,729 @@
+ * interpolate diffuse lighting while limiting sampling computation.
  */
 
 THREE.PMREMGenerator = ( function () {
 
-	var shader = getShader();
-	var camera = new THREE.OrthographicCamera( - 1, 1, 1, - 1, 0.0, 1000 );
-	var scene = new THREE.Scene();
-	var planeMesh = new THREE.Mesh( new THREE.PlaneBufferGeometry( 2, 2, 0 ), shader );
-	planeMesh.material.side = THREE.DoubleSide;
-	scene.add( planeMesh );
-	scene.add( camera );
+	var LOD_MIN = 4;
+	var LOD_MAX = 8;
+	var SIZE_MAX = Math.pow( 2, LOD_MAX );
+	// The standard deviations (radians) associated with the extra mips. These are
+	// chosen to approximate a Trowbridge-Reitz distribution function times the
+	// geometric shadowing function. These sigma values squared must match the
+	// variance #defines in cube_uv_reflection_fragment.glsl.js.
+	var EXTRA_LOD_SIGMA = [ 0.125, 0.215, 0.35, 0.446, 0.526, 0.582 ];
+	var TOTAL_LODS = LOD_MAX - LOD_MIN + 1 + EXTRA_LOD_SIGMA.length;
+	// The maximum length of the blur for loop. Smaller sigmas will use fewer
+	// samples and exit early, but not recompile the shader.
+	var MAX_SAMPLES = 20;
+	var ENCODINGS = {
+		[ THREE.LinearEncoding ]: 0,
+		[ THREE.sRGBEncoding ]: 1,
+		[ THREE.RGBEEncoding ]: 2,
+		[ THREE.RGBM7Encoding ]: 3,
+		[ THREE.RGBM16Encoding ]: 4,
+		[ THREE.RGBDEncoding ]: 5,
+		[ THREE.GammaEncoding ]: 6
+	  };
+
+	var _flatCamera = new THREE.OrthographicCamera();
+	var _blurMaterial = _getBlurShader( MAX_SAMPLES );
+
+	var { _lodPlanes, _sizeLods, _sigmas } = _createPlanes();
+	var _pingPongRenderTarget = null;
+	var _renderer = null;
+
+	// Golden Ratio
+	var PHI = ( 1 + Math.sqrt( 5 ) ) / 2;
+	var INV_PHI = 1 / PHI;
+	// Vertices of a dodecahedron (except the opposites, which represent the
+	// same axis), used as axis directions evenly spread on a sphere.
+	var _axisDirections = [
+		new THREE.Vector3( 1, 1, 1 ),
+		new THREE.Vector3( - 1, 1, 1 ),
+		new THREE.Vector3( 1, 1, - 1 ),
+		new THREE.Vector3( - 1, 1, - 1 ),
+		new THREE.Vector3( 0, PHI, INV_PHI ),
+		new THREE.Vector3( 0, PHI, - INV_PHI ),
+		new THREE.Vector3( INV_PHI, 0, PHI ),
+		new THREE.Vector3( - INV_PHI, 0, PHI ),
+		new THREE.Vector3( PHI, INV_PHI, 0 ),
+		new THREE.Vector3( - PHI, INV_PHI, 0 ) ];
+
+	var PMREMGenerator = function ( renderer ) {
+
+		_renderer = renderer;
 
-	var PMREMGenerator = function ( sourceTexture, samplesPerLevel, resolution ) {
+	};
 
-		this.sourceTexture = sourceTexture;
-		this.resolution = ( resolution !== undefined ) ? resolution : 256; // NODE: 256 is currently hard coded in the glsl code for performance reasons
-		this.samplesPerLevel = ( samplesPerLevel !== undefined ) ? samplesPerLevel : 32;
+	PMREMGenerator.prototype = {
 
-		var monotonicEncoding = ( this.sourceTexture.encoding === THREE.LinearEncoding ) ||
-			( this.sourceTexture.encoding === THREE.GammaEncoding ) || ( this.sourceTexture.encoding === THREE.sRGBEncoding );
+		constructor: PMREMGenerator,
```

Comment:
@WestLangley I'm responding to your comment here so we can have a thread about it. I found that even when I have shader debugging turned off, compiling the shaders and uploading the attributes takes as much time as running PMREM, hence I'm doing my best not to dispose of anything in case it will be reused. Would it be better to move all these private variables into the constructor as members? Or should I have a dispose method that gets rid of them all? I'm a little fuzzy on best practices in three for class architecture.

Reply:
Ultimately, I think PMREM should be handled by the renderer and transparent to the user.

In the mean time, the practice has been to call `dispose()` as-needed at the application level.

If we continue to follow that practice in the examples, then users will just copy the pattern.

Reply:
Okay, I've added a dispose method and updated the examples. Does this look like what you were hoping for? I checked that your example is now back down to the proper number of materials, geometries and textures.

Reply:
That's the idea! Thanks.